File: crates/sccgub-audit-py/sccgub_audit/chain_state.py
"""ChainStateView Protocol + JsonChainStateFixture.

Mirrors `crates/sccgub-audit/src/chain_state.rs` per PATCH_08 §B.1
+ PATCH_09 §D.4. `ChainStateView` is a `typing.Protocol` (Python's
structural-typing analog to a Rust trait); `JsonChainStateFixture`
is a `dataclass` that loads the same JSON format the Rust fixture
emits per PATCH_09 §D.6.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, List, Mapping, Optional, Protocol, Tuple


class ChainStateError(Exception):
    """Errors ChainStateView implementations may raise when the
    chain log is unreadable, corrupted, or missing required
    entries.

    Mirrors the Rust enum's variants via subclasses.
    """

    pass


class GenesisCeilingsMissing(ChainStateError):
    """The genesis ceilings record could not be located."""

    pass


class GenesisCeilingsMalformed(ChainStateError):
    """The genesis ceilings record was found but failed to
    deserialize."""

    pass


class CeilingsMissingAtHeight(ChainStateError):
    """A ChainVersionTransition referenced a height for which the
    state view has no ceilings record."""

    def __init__(self, height: int, reason: str):
        self.height = height
        self.reason = reason
        super().__init__(f"ceilings missing at height {height}: {reason}")


class ChainStateIoError(ChainStateError):
    """I/O or backend error not specific to a single height."""

    pass


# Type alias: a JSON-decoded ConstitutionalCeilings dict. Keys are
# the Rust field names exactly (per PATCH_09 §D.6).
Ceilings = Mapping[str, Any]


@dataclass(frozen=True)
class ChainVersionTransition:
    """Mirror of Rust `sccgub_types::upgrade::ChainVersionTransition`."""

    activation_height: int
    from_version: int
    to_version: int
    upgrade_spec_hash: List[int]  # 32-byte hash as JSON int array
    proposal_id: List[int]  # 32-byte hash as JSON int array

    @classmethod
    def from_json(cls, data: Mapping[str, Any]) -> "ChainVersionTransition":
        return cls(
            activation_height=int(data["activation_height"]),
            from_version=int(data["from_version"]),
            to_version=int(data["to_version"]),
            upgrade_spec_hash=list(data["upgrade_spec_hash"]),
            proposal_id=list(data["proposal_id"]),
        )


class ChainStateView(Protocol):
    """Read-only view over a chain's state required by the verifier.

    Mirrors the Rust trait. Implementations supply the three reads;
    the verifier is the only caller and uses these reads in a
    single pass.
    """

    def genesis_block_hash(self) -> List[int]:
        """The genesis block hash as a 32-byte int list."""
        ...

    def genesis_constitutional_ceilings(self) -> Ceilings:
        """The ConstitutionalCeilings as committed at genesis.

        Raises GenesisCeilingsMissing or GenesisCeilingsMalformed
        on failure.
        """
        ...

    def chain_version_history(self) -> List[ChainVersionTransition]:
        """Every ChainVersionTransition record from genesis to tip,
        ordered ascending by activation_height. Empty iff the chain
        is genesis-only."""
        ...

    def ceilings_at_height(self, height: int) -> Ceilings:
        """The ceilings record as committed at block `height`.

        Raises CeilingsMissingAtHeight if the height has no
        ceilings record in this view.
        """
        ...


# ─── JsonChainStateFixture ────────────────────────────────────────

@dataclass
class JsonChainStateFixture:
    """A ChainStateView backed by an in-memory JSON-shaped fixture.

    Designed for tests, the CLI v1 `--chain-state <path>` mode, and
    the cross-language conformance harness (PATCH_09 §E).

    Reads the **identical JSON fixture format** the Rust port
    produces, per PATCH_09 §D.6.
    """

    genesis_block_hash: List[int]
    genesis_ceilings: Ceilings
    chain_version_history_list: List[ChainVersionTransition]
    ceilings_by_height: List[Tuple[int, Ceilings]]

    def genesis_block_hash_(self) -> List[int]:
        return self.genesis_block_hash

    def genesis_constitutional_ceilings(self) -> Ceilings:
        return self.genesis_ceilings

    def chain_version_history(self) -> List[ChainVersionTransition]:
        return list(self.chain_version_history_list)

    def ceilings_at_height(self, height: int) -> Ceilings:
        for h, c in self.ceilings_by_height:
            if h == height:
                return c
        raise CeilingsMissingAtHeight(
            height,
            f"no ceilings record in fixture for height {height}",
        )

    # NOTE: Python's Protocol is structural — `genesis_block_hash`
    # is a method on the Protocol but a Python attribute access
    # would conflict with the dataclass field of the same name.
    # We expose `genesis_block_hash_` for the verifier's use
    # internally; the verifier prefers the field directly when
    # a JsonChainStateFixture is passed.

    @classmethod
    def from_json(cls, data: Mapping[str, Any]) -> "JsonChainStateFixture":
        """Load from a JSON-decoded dict.

        The JSON shape mirrors Rust serde:

            {
                "genesis_block_hash": [u8; 32 ints],
                "genesis_ceilings": { <ceiling_field>: <int>, ... },
                "chain_version_history": [ <transition>, ... ],
                "ceilings_by_height": [ [<height>, <ceilings>], ... ]
            }
        """
        return cls(
            genesis_block_hash=list(data["genesis_block_hash"]),
            genesis_ceilings=dict(data["genesis_ceilings"]),
            chain_version_history_list=[
                ChainVersionTransition.from_json(t)
                for t in data["chain_version_history"]
            ],
            ceilings_by_height=[
                (int(pair[0]), dict(pair[1]))
                for pair in data["ceilings_by_height"]
            ],
        )


def load_fixture_from_json(path_or_text: Any) -> JsonChainStateFixture:
    """Load a JsonChainStateFixture from either a filesystem path
    or a JSON string. CLI helper.
    """
    if hasattr(path_or_text, "read"):
        text = path_or_text.read()
    elif isinstance(path_or_text, str) and "\n" in path_or_text:
        # Looks like JSON content already (multi-line string).
        text = path_or_text
    elif isinstance(path_or_text, str):
        # Treat as file path.
        with open(path_or_text, "r", encoding="utf-8") as fh:
            text = fh.read()
    else:
        # Path-like object.
        with open(path_or_text, "r", encoding="utf-8") as fh:
            text = fh.read()
    data = json.loads(text)
    return JsonChainStateFixture.from_json(data)


def genesis_preserved_fixture(
    genesis_block_hash: List[int],
    genesis_ceilings: Ceilings,
    history: List[ChainVersionTransition],
) -> JsonChainStateFixture:
    """Construct a "happy path" fixture where every height retains
    the genesis ceilings exactly. Mirrors Rust
    `JsonChainStateFixture::genesis_preserved`.
    """
    by_height: List[Tuple[int, Ceilings]] = []
    for t in history:
        if t.activation_height > 0:
            by_height.append((t.activation_height - 1, dict(genesis_ceilings)))
        by_height.append((t.activation_height, dict(genesis_ceilings)))
    return JsonChainStateFixture(
        genesis_block_hash=list(genesis_block_hash),
        genesis_ceilings=dict(genesis_ceilings),
        chain_version_history_list=list(history),
        ceilings_by_height=by_height,
    )

File: crates/sccgub-audit-ts/src/chainState.ts
/**
 * ChainStateView interface + JsonChainStateFixture.
 *
 * Mirrors `crates/sccgub-audit/src/chain_state.rs` per PATCH_08 §B.1
 * + PATCH_09 §C.4. `ChainStateView` is a TypeScript interface
 * (structural-typing analog to a Rust trait); `JsonChainStateFixture`
 * is a class that loads the same JSON format the Rust fixture emits
 * per PATCH_09 §E.
 */

import { readFileSync } from "node:fs";

/** Errors ChainStateView implementations may raise. */
export class ChainStateError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ChainStateError";
  }
}

/** The genesis ceilings record could not be located. */
export class GenesisCeilingsMissing extends ChainStateError {
  constructor(message: string) {
    super(message);
    this.name = "GenesisCeilingsMissing";
  }
}

/** The genesis ceilings record was found but failed to deserialize. */
export class GenesisCeilingsMalformed extends ChainStateError {
  constructor(message: string) {
    super(message);
    this.name = "GenesisCeilingsMalformed";
  }
}

/**
 * A ChainVersionTransition referenced a height for which the state
 * view has no ceilings record.
 */
export class CeilingsMissingAtHeight extends ChainStateError {
  readonly height: bigint;
  readonly reason: string;
  constructor(height: bigint, reason: string) {
    super(`ceilings missing at height ${height}: ${reason}`);
    this.name = "CeilingsMissingAtHeight";
    this.height = height;
    this.reason = reason;
  }
}

/** I/O or backend error not specific to a single height. */
export class ChainStateIoError extends ChainStateError {
  constructor(message: string) {
    super(message);
    this.name = "ChainStateIoError";
  }
}

/**
 * Type alias: a JSON-decoded ConstitutionalCeilings dict. Keys are
 * the Rust field names exactly (per PATCH_09 §E).
 */
export type Ceilings = Readonly<Record<string, unknown>>;

/** Mirror of Rust `sccgub_types::upgrade::ChainVersionTransition`. */
export interface ChainVersionTransition {
  readonly activation_height: bigint;
  readonly from_version: number;
  readonly to_version: number;
  readonly upgrade_spec_hash: readonly number[]; // 32 bytes
  readonly proposal_id: readonly number[]; // 32 bytes
}

/**
 * Construct a {@link ChainVersionTransition} from a JSON-decoded
 * mapping. Coerces fields to their canonical types.
 */
export function chainVersionTransitionFromJson(
  data: Readonly<Record<string, unknown>>,
): ChainVersionTransition {
  return {
    activation_height: toBigInt(data["activation_height"]),
    from_version: Number(data["from_version"]),
    to_version: Number(data["to_version"]),
    upgrade_spec_hash: toByteArray(data["upgrade_spec_hash"]),
    proposal_id: toByteArray(data["proposal_id"]),
  };
}

/**
 * Read-only view over a chain's state required by the verifier.
 *
 * Mirrors the Rust trait. Implementations supply the four reads;
 * the verifier is the only caller and uses these reads in a single
 * pass.
 */
export interface ChainStateView {
  /** The genesis block hash as a 32-byte number array. */
  genesisBlockHash(): readonly number[];

  /**
   * The ConstitutionalCeilings as committed at genesis.
   * @throws {@link GenesisCeilingsMissing} or {@link GenesisCeilingsMalformed}.
   */
  genesisConstitutionalCeilings(): Ceilings;

  /**
   * Every ChainVersionTransition record from genesis to tip,
   * ordered ascending by activation_height. Empty iff the chain is
   * genesis-only.
   */
  chainVersionHistory(): readonly ChainVersionTransition[];

  /**
   * The ceilings record as committed at block `height`.
   * @throws {@link CeilingsMissingAtHeight} if absent.
   */
  ceilingsAtHeight(height: bigint): Ceilings;
}

/**
 * A ChainStateView backed by an in-memory JSON-shaped fixture.
 *
 * Designed for tests, the CLI v1 `--chain-state <path>` mode, and
 * the cross-language conformance harness (PATCH_09 §E).
 *
 * Reads the **identical JSON fixture format** the Rust port produces,
 * per PATCH_09 §E.
 */
export class JsonChainStateFixture implements ChainStateView {
  readonly genesis_block_hash: readonly number[];
  readonly genesis_ceilings: Ceilings;
  readonly chain_version_history: readonly ChainVersionTransition[];
  readonly ceilings_by_height: ReadonlyArray<readonly [bigint, Ceilings]>;

  constructor(
    genesisBlockHash: readonly number[],
    genesisCeilings: Ceilings,
    chainVersionHistory: readonly ChainVersionTransition[],
    ceilingsByHeight: ReadonlyArray<readonly [bigint, Ceilings]>,
  ) {
    this.genesis_block_hash = genesisBlockHash;
    this.genesis_ceilings = genesisCeilings;
    this.chain_version_history = chainVersionHistory;
    this.ceilings_by_height = ceilingsByHeight;
  }

  genesisBlockHash(): readonly number[] {
    return this.genesis_block_hash;
  }

  genesisConstitutionalCeilings(): Ceilings {
    return this.genesis_ceilings;
  }

  chainVersionHistory(): readonly ChainVersionTransition[] {
    return this.chain_version_history;
  }

  ceilingsAtHeight(height: bigint): Ceilings {
    for (const [h, c] of this.ceilings_by_height) {
      if (h === height) {
        return c;
      }
    }
    throw new CeilingsMissingAtHeight(
      height,
      `no ceilings record in fixture for height ${height}`,
    );
  }

  /** Load from a JSON-decoded value matching the Rust serde shape. */
  static fromJson(data: Readonly<Record<string, unknown>>): JsonChainStateFixture {
    const ceilingsByHeightRaw = data["ceilings_by_height"];
    if (!Array.isArray(ceilingsByHeightRaw)) {
      throw new TypeError(
        "fixture.ceilings_by_height must be an array of [height, ceilings] pairs",
      );
    }
    const ceilingsByHeight: Array<readonly [bigint, Ceilings]> = [];
    for (const pair of ceilingsByHeightRaw) {
      if (!Array.isArray(pair) || pair.length !== 2) {
        throw new TypeError(
          "fixture.ceilings_by_height entries must be 2-element arrays",
        );
      }
      ceilingsByHeight.push([
        toBigInt(pair[0]),
        pair[1] as Ceilings,
      ] as const);
    }
    const historyRaw = data["chain_version_history"];
    if (!Array.isArray(historyRaw)) {
      throw new TypeError("fixture.chain_version_history must be an array");
    }
    const history = historyRaw.map((t) =>
      chainVersionTransitionFromJson(t as Record<string, unknown>),
    );
    return new JsonChainStateFixture(
      toByteArray(data["genesis_block_hash"]),
      data["genesis_ceilings"] as Ceilings,
      history,
      ceilingsByHeight,
    );
  }
}

/**
 * Load a {@link JsonChainStateFixture} from either a filesystem
 * path or a JSON string. CLI helper.
 *
 * Uses {@link parseJsonPreservingBigInts} so that integer literals
 * larger than `Number.MAX_SAFE_INTEGER` (2^53 − 1) — which include
 * `min_effective_fee_floor` (10^16) and i128-typed ceilings — are
 * preserved as `bigint` rather than truncated to IEEE-754 doubles.
 * Without this, drift on a large-value field is invisible because
 * `1e16` and `1e16 - 1` round to the same Number.
 */
export function loadFixtureFromJson(pathOrText: string): JsonChainStateFixture {
  let text: string;
  if (pathOrText.includes("\n") && pathOrText.trimStart().startsWith("{")) {
    // Looks like JSON content already (multi-line string starting with `{`).
    text = pathOrText;
  } else {
    // Treat as file path.
    text = readFileSync(pathOrText, "utf8");
  }
  const data = parseJsonPreservingBigInts(text) as Record<string, unknown>;
  return JsonChainStateFixture.fromJson(data);
}

/**
 * Parse JSON, returning `bigint` for integer literals outside the
 * IEEE-754 safe-integer range and `number` otherwise.
 *
 * Implementation: pre-process the text by wrapping unsafe integer
 * literals in a sentinel string, then parse via stdlib `JSON.parse`
 * with a reviver that converts the sentinel back to `bigint`. This
 * keeps the function pure-stdlib (PATCH_09 §D.3) while preserving
 * value-correctness for i128-typed ceiling fields per PATCH_09 §C.
 *
 * Caveat: the regex matches integer literals in JSON value position
 * only (after `:`, `[`, or `,`). It does NOT match inside object
 * keys (always strings in JSON) or fractional numbers (which would
 * be Number anyway). The pattern is conservative.
 */
export function parseJsonPreservingBigInts(text: string): unknown {
  const SAFE_DIGITS = String(Number.MAX_SAFE_INTEGER).length; // 16
  const SENTINEL = "__sccgub_bigint__";
  const wrapped = text.replace(
    /([:\[,]\s*)(-?\d+)(?=\s*[,\]}])/g,
    (match, prefix: string, num: string): string => {
      const abs = num.startsWith("-") ? num.slice(1) : num;
      // Fast path: clearly safe-range. Leave as-is so it parses to Number.
      if (abs.length < SAFE_DIGITS) return match;
      // Borderline (16 digits): compare lexicographically against the
      // string form of MAX_SAFE_INTEGER. Equal-length strings compare
      // numerically when both are positive integer strings.
      if (
        abs.length === SAFE_DIGITS &&
        abs <= String(Number.MAX_SAFE_INTEGER)
      ) {
        return match;
      }
      return `${prefix}"${SENTINEL}${num}"`;
    },
  );
  return JSON.parse(wrapped, (_key: string, value: unknown): unknown => {
    if (typeof value === "string" && value.startsWith(SENTINEL)) {
      return BigInt(value.slice(SENTINEL.length));
    }
    return value;
  });
}

/**
 * Construct a "happy path" fixture where every height retains the
 * genesis ceilings exactly. Mirrors Rust
 * `JsonChainStateFixture::genesis_preserved` and Python
 * `genesis_preserved_fixture`.
 */
export function genesisPreservedFixture(
  genesisBlockHash: readonly number[],
  genesisCeilings: Ceilings,
  history: readonly ChainVersionTransition[],
): JsonChainStateFixture {
  const byHeight: Array<readonly [bigint, Ceilings]> = [];
  for (const t of history) {
    if (t.activation_height > 0n) {
      byHeight.push([t.activation_height - 1n, { ...genesisCeilings }] as const);
    }
    byHeight.push([t.activation_height, { ...genesisCeilings }] as const);
  }
  return new JsonChainStateFixture(
    [...genesisBlockHash],
    { ...genesisCeilings },
    [...history],
    byHeight,
  );
}

// ─── Internal coercion helpers ────────────────────────────────────

function toBigInt(v: unknown): bigint {
  if (typeof v === "bigint") return v;
  if (typeof v === "number") {
    if (!Number.isInteger(v)) {
      throw new TypeError(`expected integer, got ${v}`);
    }
    return BigInt(v);
  }
  if (typeof v === "string") return BigInt(v);
  throw new TypeError(`cannot convert ${JSON.stringify(v)} to bigint`);
}

function toByteArray(v: unknown): readonly number[] {
  if (!Array.isArray(v)) {
    throw new TypeError(`expected byte array, got ${typeof v}`);
  }
  return v.map((x) => {
    if (typeof x !== "number" || !Number.isInteger(x) || x < 0 || x > 255) {
      throw new TypeError(`byte-array element out of range: ${JSON.stringify(x)}`);
    }
    return x;
  });
}
